### PR TITLE
miscellaneous features

### PR DESCRIPTION
**Pull request by Dudemanguy**
Handful of small features I'd thought I'd implement. #12131 was originally a part of this but that became surprisingly complicated so I split it out.

### CODE REVIEW

**Review by christoph-heinrich — CHANGES_REQUESTED** [review]
`playlist-path` is a nice feature, and is the only thing I've tested.

File: DOCS/interface-changes.rst [hunk 1]
```diff
@@ -35,6 +35,9 @@ Interface changes
     - include `--hdr-peak-percentile` in the `gpu-hq` profile
     - change `--audiotrack-pcm-float` default from `no` to `yes`
     - change type of `--sub-pos` to float
+    - add `--term-timetotal` option
```
[the remark this answers is not in the document]
Comment (Dudemanguy) [reply DOCS/interface-changes.rst:38]: That was what I originally called it, but then changed it and somehow missed this line heh.

File: video/filter/vf_vapoursynth.c [hunk 2]
```diff
@@ -89,6 +89,7 @@ struct priv {
     int64_t frames_sent;        // total nr. of frames ever added to input queue
     bool initializing;          // filters are being built
     bool in_node_active;        // node might still be called
+    bool display_res;           // resolution of the display was obtained
```
[the remark this answers is not in the document]
Comment (Dudemanguy) [reply video/filter/vf_vapoursynth.c:92]: Agreed

File: player/osd.c [hunk 3]
```diff
@@ -195,9 +195,10 @@ static char *get_term_status_msg(struct MPContext *mpctx)
     saddf(&line, ": ");
 
     // Playback position
+    double speed = opts->term_remaining_playtime ? opts->playback_speed : 1;
```
Comment (sfan5) [inline player/osd.c:198]: `mp_property_playtime_remaining()` uses `mpctx->video_speed` instead

File: filters/filter.h [hunk 4]
```diff
@@ -398,6 +398,7 @@ struct mp_stream_info {
     void *priv; // for use by whoever implements the callbacks
 
     double (*get_display_fps)(struct mp_stream_info *i);
+    void   (*get_display_res)(struct mp_stream_info *i, void *res);
```
Comment (sfan5) [inline filters/filter.h:401]: why not `int64_t *`?

File: video/filter/vf_vapoursynth.c [hunk 5]
```diff
@@ -312,6 +313,21 @@ static void vf_vapoursynth_process(struct mp_filter *f)
...
+    // window to be ready.
+    if (!p->got_display_res) {
+        struct mp_stream_info *info = mp_filter_find_stream_info(p->f);
+        int64_t res[2] = {0};
+        if (info && info->get_display_res)
+            info->get_display_res(info, res);
+        if (res[0] > 0) {
+            p->got_display_res = true;
+            VSMap *vars = p->vsapi->createMap();
+            p->vsapi->propSetIntArray(vars, "display_res", res, 2);
```
Comment (sfan5) [inline video/filter/vf_vapoursynth.c:326]: document this somewhere?

File: filters/filter.h (hunk 4 above)
Comment (Dudemanguy) [reply filters/filter.h:401]: I don't think this can be in64_t * without allocating memory? That's why I went with void pointers.

File: video/filter/vf_vapoursynth.c (hunk 5 above)
Comment (Dudemanguy) [reply video/filter/vf_vapoursynth.c:326]: In the manpages now.

File: filters/filter.h (hunk 4 above)
Comment (christoph-heinrich) [reply filters/filter.h:401]: Why would a int64_t pointer require an allocation when a void pointer can do it without? The void pointer still has to point somewhere, and at the place where it's called there is a int64_t array. that ends up getting pointed to. I don't see how the type changes anything here regarding memory.

Comment (Dudemanguy) [reply filters/filter.h:401]: Maybe I was just returning the wrong address. Let me try again.

Comment (Dudemanguy) [reply filters/filter.h:401]: Yeah, not sure how to do it like this. Maybe I'm just bad at C but the in64_t array is always local and I'm not sure how you would go about returning it. Plus there's also an additional conversion step you would have to do because VOCTRL_GET_DISPLAY_RES requires int[2] not int64_t[2].

Comment (christoph-heinrich) [reply filters/filter.h:401]: Then the array we pass in also needs to be int[2] and then converted into int64_t[2]. Making the pointer void doesn't change the location of the values. Alternatively we could also change VOCTRL_GET_DISPLAY_RES to int64_t[2], not sure if that's viable.

Comment (Dudemanguy) [reply filters/filter.h:401]: Yeah it was returning garbage values. Should be fixed now.

>VOCTRL_GET_DISPLAY_RES to int64_t[2]

Not worth it imo.

File: DOCS/man/vf.rst [hunk 6]
```diff
@@ -583,6 +583,10 @@ Available mpv-only filters are:
     ``display_fps``
         Refresh rate of the current display. Note that this value can be 0.
 
+    ``display_res``
+        Resolution of the current display. Note that this will not react
+        to monitor changes and may not work with single image streams.
+
```
Comment (hooke007) [inline DOCS/man/vf.rst:589]: What's format of this var?
`[aa, bb]` or anything else?

Comment (Dudemanguy) [reply DOCS/man/vf.rst:589]: It's an int64_t array. I use vapoursynth so I don't know what that actual looks like in practical usage. The types of the other variables aren't documented at all so I assume it doesn't really matter.

Comment (hooke007) [reply DOCS/man/vf.rst:589]: I mean if users want to use it in vs. They know how to use it quickily.
Here is a example:
```
if display_res.width > 1920 :
    .....
if display_res.height > 1080 :
    .....
```
But now apparently they are not sure about the type of this var.

Comment (Dudemanguy) [reply DOCS/man/vf.rst:589]: Fair point. I presume the syntax is more like `if display_res[0] > 1920` (or however you subselect an array). I'll elaborate a bit here.

File: filters/filter.h (hunk 4 above)
Comment (christoph-heinrich) [reply filters/filter.h:401]: Agree

**Comment by hooke007** [comment]
I had the another question about vf_vapoursynth.
Currently mpv can only show vpy's error msg.
```python
# test.vpy
raise Exception("vs is disabled")
```
Snipaste_2023-08-12_02-20-57


Is there any way to make mpv's console output vpy's `print()`?

Here is a minimal test example.
`--vf=vapoursynth="test.vpy"`
```python
# test.vpy
print("vs is working")
video_in.set_output()


```

**Comment by Dudemanguy** [comment]
Do mean as part of mpv's logger? I don't know. The raiseException example you mention gets printed because it throws an error while mpv is parsing the script so it gets printed.

**Comment by hooke007** [comment]
Well. I write vpy for mpv. It's really painful because there is no direct way to check var's state.
If mpv parsed the script successfully, nothing would be printed. Though some var were changed out of expectation. It's difficult to debug.

**Comment by Dudemanguy** [comment]
So on the vapoursynth front, I added a couple of commits to X11 and Wayland that make them return these `display-` properties earlier and thus make the variables actually useful. You can test it out with:
```
#!/usr/bin/env python
video_in.text.Text(f'{display_res[0]} x {display_res[1]}, {display_fps}Hz').set_output()
````
It should print values now.

File: filters/filter.h (hunk 4 above)
Comment (sfan5) [reply filters/filter.h:401]: ah sorry I misread, this can be `int *` however

nothing to do with allocations, it's just the type that's actually being used.

Comment (Dudemanguy) [reply filters/filter.h]: I just needed to make it static.

**Comment by Dudemanguy** [comment]
I dropped the X11 commit because it ended up being a big mess with how it handles all this initial geometry stuff. I'll do a separate PR for that later. So the above script won't display the display width/height on x11 (it'll be 0x0). The fps works though. And it works on wayland since that change was simple.

File: filters/f_output_chain.c [hunk 7]
```diff
@@ -357,6 +357,15 @@ static double get_display_fps(struct mp_stream_info *i)
     return res;
 }
 
+static int *get_display_res(struct mp_stream_info *i)
+{
+    struct chain *p = i->priv;
+    static int res[2] = {0};
```
Comment (sfan5) [inline filters/f_output_chain.c:363]: what no
static locals are to be avoided at all costs

Comment (sfan5) [reply filters/f_output_chain.c:363]: to be clear: what I meant from the beginning was that the signature should be `static void get_display_res(struct mp_stream_info *i, int *dst)`

Comment (Dudemanguy) [reply filters/f_output_chain.c:363]: oh

**Review by sfan5 — APPROVED** [review]
